PDFs should be downloaded in lockdown mode

robert-jenner opened the pull request:
#### 5dc8c9e2659149429af495713cb8cf8a62ef4dce
<pre>
PDFs should be downloaded in lockdown mode
<a href="https://bugs.webkit.org/show_bug.cgi?id=271003">https://bugs.webkit.org/show_bug.cgi?id=271003</a>
<a href="https://rdar.apple.com/124636755">rdar://124636755</a>

Reviewed by Chris Dumez.

This functionality was first introduced in <a href="https://bugs.webkit.org/show_bug.cgi?id=237245">https://bugs.webkit.org/show_bug.cgi?id=237245</a>, but
was disabled for macOS in <a href="https://bugs.webkit.org/show_bug.cgi?id=241741.">https://bugs.webkit.org/show_bug.cgi?id=241741.</a> This patch is to re-enable it.

This functionality should be covered by WKDownload.LockdownModePDF. Which is also fixed in this change
such that it fails whenever the conditional doesn&apos;t exist.

* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::decidePolicyForResponseShared):
* Tools/TestWebKitAPI/Tests/WebKitCocoa/Download.mm:

Originally-landed-as: 272448.740@safari-7618-branch (c369f5018e91). <a href="https://rdar.apple.com/128088854">rdar://128088854</a>
Canonical link: <a href="https://commits.webkit.org/278919@main">https://commits.webkit.org/278919@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/5a382a7bfc3733c36af97452af7b54ced92cec2c

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/51872 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/31184 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/4236 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/55140 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/2564 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/54175 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/37559 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/2274 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/42190 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/53971 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/28797 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/44744 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/23329 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/26074 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/2006 "Passed tests") | [⏳ 🛠 wpe-skia ](https://ews-build.webkit.org/#/builders/WPE-Skia-Build-EWS "Waiting in queue, processing has not started yet") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/48028 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/2098 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/56731 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/26993 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/1977 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/49603 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/28230 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/44855 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/48846 "Passed tests") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/29127 "Built successfully") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/7593 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/27967 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

pascoej commented on Source/WebKit/UIProcess/WebPageProxy.cpp:
```diff
@@ -7365,6 +7365,24 @@ void WebPageProxy::decidePolicyForResponseShared(Ref<WebProcessProxy>&& process,
+        bool shouldForceDownload = [&] {
+            if (policyAction != PolicyAction::Use || process->lockdownMode() != WebProcessProxy::LockdownMode::Enabled)
+                return false;
+            if (MIMETypeRegistry::isPDFOrPostScriptMIMEType(navigationResponse->response().mimeType()))
```
```suggestion
            if (MIMETypeRegistry:: isPDFMIMEType(navigationResponse->response().mimeType()))
```

pascoej commented on Source/WebKit/UIProcess/WebPageProxy.cpp:
```diff
@@ -7365,6 +7365,24 @@ void WebPageProxy::decidePolicyForResponseShared(Ref<WebProcessProxy>&& process,
+        if (shouldForceDownload)
+            policyAction = PolicyAction::Download;
+
+        Ref sender = PolicyDecisionSender::create(WTFMove(completionHandler));
```
```suggestion
        receivedNavigationResponsePolicyDecision(policyAction, navigation.get(), request, WTFMove(navigationResponse), WTFMove(completionHandler));
```

pascoej commented on Source/WebKit/UIProcess/WebPageProxy.cpp:
```diff
@@ -7365,6 +7365,24 @@ void WebPageProxy::decidePolicyForResponseShared(Ref<WebProcessProxy>&& process,
+            policyAction = PolicyAction::Download;
+
+        Ref sender = PolicyDecisionSender::create(WTFMove(completionHandler));
+        receivedPolicyDecision(policyAction, navigation.get(), nullptr, WTFMove(navigationResponse), WTFMove(sender), WillContinueLoadInNewProcess::No, std::nullopt);
```
```suggestion
```

pascoej committed:
PDFs should be downloaded in lockdown mode
https://bugs.webkit.org/show_bug.cgi?id=271003
rdar://124636755

Reviewed by Chris Dumez.

This functionality was first introduced in https://bugs.webkit.org/show_bug.cgi?id=237245, but
was disabled for macOS in https://bugs.webkit.org/show_bug.cgi?id=241741. This patch is to re-enable it.

This functionality should be covered by WKDownload.LockdownModePDF. Which is also fixed in this change
such that it fails whenever the conditional doesn't exist.

* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::decidePolicyForResponseShared):
* Tools/TestWebKitAPI/Tests/WebKitCocoa/Download.mm:

Originally-landed-as: 272448.740@safari-7618-branch (c369f5018e91). rdar://128088854
Canonical link: https://commits.webkit.org/278919@main
